astropy/units/core.py

```diff
@@ -846,16 +858,17 @@ def __rtruediv__(self, m):
             return NotImplemented
 
     def __mul__(self, m):
-        if isinstance(m, (bytes, str)):
-            m = Unit(m)
-
         if isinstance(m, UnitBase):
             if m.is_unity():
                 return self
             elif self.is_unity():
                 return m
             return CompositeUnit(1, [self, m], [1, 1], _error_check=False)
 
+        if isinstance(m, (bytes, str)):
+            self._warn_about_operation_with_deprecated_type("products", m)
+            return self * Unit(m)
```

Comment:
Is there a good reason to recurse here? If the `self._warn_about_operation_with_deprecated_type("products", m)` is added to deleted lines 849-851, wouldn't this be unnecessary?

Reply:
With this change the very common case of `m` being an `UnitBase` instance can be fully handled without having to check `isinstance(m, (bytes, str))` at all, which speeds it up (a little bit). The tradeoff is that if `m` is a `str` or a `bytes` then the updated code might be a bit slower, but we are deprecating that code path, so its performance is not important.

Reply:
OK, I see the logic now, doing this both for `__mul__` and `__truediv__`